### PR TITLE
Add retry support to every request

### PR DESCRIPTION
**Pull request by houshengbo**
This PR added a retry function applicable to all the requests sent
to openwhisk services. Since openwhisk supports HA for controllers,
which can pick up requests and handles them in a parallel fashion,
we need to avoid the race condition.

### CODE REVIEW

File: whisk/client.go [hunk 1]
```diff
@@ -49,6 +50,8 @@ const (
     ExitWithErrorOnTimeout = true
     ExitWithSuccessOnTimeout = false
     DEFAULT_HTTP_TIMEOUT = 30
+    DEFAULT_ATTEMPTS = 5
+    DEFAULT_INTERVAL = 2 * time.Second
```
Comment (mrutkows) [inline whisk/client.go]: Curious, why you selected 2 secs? I have seen a progressive approach where waiting 1 sec. on first iteration, 2 seconds on 2nd iteration, 4 sec. on 3rd iteration, and so forth. Really, my hope is that the HTTP request turnaround/round-trip time itself allows for success on a subsequent call. Do we really need to wait 2 seconds in addition to that?

Comment (mrutkows) [reply whisk/client.go]: Also, 3 attempts seems to be an industry standard...

File: whisk/client.go [hunk 2]
```diff
@@ -316,98 +319,118 @@ func BodyTruncator(body io.ReadCloser) (string, io.ReadCloser, error) {
     return string(data), reload, nil
 }
 
+func retry(attempts int, sleep time.Duration, callback func() (*http.Response, error)) (*http.Response, error)  {
+    var err error
+    var resp *http.Response
+    for i := 0; ; i++ {
+        resp, err = callback()
```
Comment (mrutkows) [inline whisk/client.go]: @houshengbo do we wish to retry on all errors?  It seems that some are "soft" errors that deserve retry (like concurrency) and others are "hard" errors that will not produce different results on a retry (for example, authorization failures).

Comment (rabbah) [reply whisk/client.go]: retry shouldn't be applied blindly - a conflict (409) is an indication that something is wrong and automatically retrying may not have the desired affect in some cases.

Comment (starpit) [reply whisk/client.go]: the logic is actually a bit subtle. i have this in a branch of the openwhisk npm -- where i've needed to use retries for a while, now.

the following status codes should never be retried:
```
statusCode === 400 || statusCode === 401 || statusCode === 404 || statusCode === 409 || statusCode === 410 || statusCode === 500
```

if the statusCode is 502... then we probably also shouldn't retry.

Comment (mrutkows) [reply whisk/client.go]: we could retry on non-HTTP error code of 153, where we know we should retry...
```
Error code: 153: Concurrent modification to resource detected.
```

Comment (rabbah) [reply whisk/client.go]: 153 is 409 % 256 which is the wrapping the client does to report the status code on the terminal. 

a blind retry on conflict just doesn't make sense to me: what does it mean for example to retry rule enable/disable if the concurrent operation deleted the rule, for example.

Comment (houshengbo) [reply whisk/client.go]: What about checking both of the "code 409" and error message "Concurrent modification to resource detected"? Can this be limited to the case of Concurrent modification only?

**Comment by rabbah** [comment]
I’m really confused what the goals are. The 409s you’re seeing are specific to your test cases. Why not add retries in the tests for wskdeploy? 

Nicks suggestion for which codes to retry makes sense. Beyond that it’s dubious what you’re trying to get out of a retry.

**Comment by houshengbo** [comment]
It is not specific to my test cases. It will happen to all the situations, if multiple requests are sent to openwhisk in a very short time.
As long as the user using client side code will bump into this issue, if requests are sent fast enough.

**Comment by rabbah** [comment]
That’s right: the client which is the user of this library should decide on the retry policy. Not this go library itself. What does it mean to retry a conflict which inherently means an invariant might have been violated. The compensation and resolution will be specific to a scenario. Like the test cases.

**Comment by rabbah** [comment]
Let’s focus on a specific example where you first observed the 409. 

You disable a rule. 
Then you delete the rule. 

Say the delete succeeds. The disable will fail with a 404. Does it make sense to retry in this case? You’ll have to adjust your test case anyway. 

If you retry the delete because the disable won the race, you’ll get the test outcome you want. However since this is a race already you have to adjust the test case either way. 

So can we work through a couple of example so that we can clarify what you’re trying to accomplish?

**Comment by mdeuser** [comment]
I agree with @rabbah - let's get specific on these HA/race-condition issues and then decide on the best practice way to handle these.  developers using the REST API will need to know how to deal with these issues as well.

**Comment by dubeejw** [comment]
@houshengbo, what's the status of this PR?

**Pull request closed without merging**